docs/hands-on-learning/advanced/module-3-traits.md

```diff
@@ -0,0 +1,82 @@
+---
+sidebar_label: "Module 3: Traits"
+description: How to use traits to match semantically related elements.
+---
+
+# Module 3: Traits
+
+Traits are a powerful abstraction that allow you to define higher-level semantic groupings in OpenRewrite’s LSTs (Lossless Syntax Trees). They let you build reusable logic for elements that are semantically similar but structurally different. Instead of embedding utility logic in unrelated classes or expanding the core LST APIs, traits act as opt-in behavior layers. This keeps your recipes modular, discoverable, and semantically rich.
+
+For example, if you want to operate on all classes annotated with `@Bean`, regardless of their structure or placement, traits allow you to define a matcher that groups those together. We will look at this example in this module.
+
+## Exercise 3a: Explore a recipe that uses Traits
+
+In this exercise, you'll use the `Annotated.Matcher` trait that the [`FindSpringBeans`](https://github.com/moderneinc/rewrite-recipe-starter/blob/main/src/main/java/com/yourorg/FindSpringBeans.java) recipe uses to identify classes annotated with `@Bean` and marks them using `SearchResult.found`.
+
+### Goals for this exercise
+
+* Understand the purpose of traits and when to use them.
+* See how the `Annotated.Matcher` trait is used to build reusable match logic.
+
+### Steps
+
+1. With the [`rewrite-recipe-starter`](https://github.com/moderneinc/rewrite-recipe-starter) still open in IntelliJ, open the `FindSpringBeans` recipe.
+   * You can find this recipe in [src/main/java/com/yourorg/FindSpringBeans.java](https://github.com/moderneinc/rewrite-recipe-starter/blob/main/src/test/java/com/yourorg/FindSpringBeans.java).
```

Comment:
The FindSpringBeans links results in 404s for me. I'm wondering if something went wrong as it seems like most of the github links are 404s. Perhaps there's some permission that's gone wrong or something?

Reply:
Again, dependent on the PRs being merged in the rewrite starter repo.